From abb8638b7403ccbfe95cd327b2e96eaae3f945f4 Mon Sep 17 00:00:00 2001
From: Avula Sri Charan <quic_asrichar@quicinc.com>
Date: Wed, 16 Oct 2024 15:10:42 +0530
Subject: QSDK: PLATFORM: wifi: ath11k: Validate data size received
 from firmware_request_nowarn()

When multiple hardware request the file in a short period
of time from firmware through firmware_request_nowarn(),
sometime we receive file with size zero if requested file
is not present at firmware path.

For example, we are requesting for fw.dat file. There
are 2 hardware (hw1 and hw2) that request for fw.dat file
at the same time. hw1 sends request through
firmware_request_nowarn() to get fw.dat file. This fw.dat
file is not present at firmware path. While requesting
the fw.dat file, first check fw_priv is present for
fw.dat file or not  through __lookup_fw_priv(), if fw_priv
is not present then allocate fw_priv to it in
alloc_lookup_fw_priv(). At the same time hw2 requests for
fw.dat file, now __lookup_fw_priv() finds fw_priv pointer
that is allocated for hw1, and proceed as a batch request
shares the same struct fw_priv and in
alloc_lookup_fw_priv() which returns success to driver
for hw2 before hw1 comes out of fw_load_sysfs_fallback().

For hw1, in fw_load_sysfs_fallback() as the fw.dat file
is not present at firmware path the fw_priv->data is
null fw_load_sysfs_fallback() returns -ENOMEM. Upon
finding this out that fw.dat file is not present at
firmware path _request_firmware() aborts batch requests
(fw_abort_batch_reqs()) and releases firmware
(release_firmware()). But for hw2, driver gets firmware
that has zero size. This may cause boot-up failure.

Add a check to avoid fetching of zero size file.

This issue can be seen  when module reboot with mlo_capable=0
cold_boot_cal=0.

Signed-off-by: Aaradhana Sahu <quic_aarasahu@quicinc.com>
Signed-off-by: Avula Sri Charan <quic_asrichar@quicinc.com>
---
 drivers/net/wireless/ath/ath11k/core.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/drivers/net/wireless/ath/ath11k/core.c b/drivers/net/wireless/ath/ath11k/core.c
index 4d18782f..af253ae4 100644
--- a/drivers/net/wireless/ath/ath11k/core.c
+++ b/drivers/net/wireless/ath/ath11k/core.c
@@ -1356,6 +1356,9 @@ const struct firmware *ath11k_core_firmware_request(struct ath11k_base *ab,
 	if (ret)
 		return ERR_PTR(ret);
 
+	if (fw && !fw->size)
+		return ERR_PTR(-ENOENT);
+
 	ath11k_dbg(ab, ATH11K_DBG_BOOT, "firmware request %s size %zu\n",
 		   path, fw->size);
 
-- 
2.34.1

